File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/Color.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import android.content.Context;

import com.urbanairship.android.layout.util.ResourceUtils;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonList;
import com.urbanairship.json.JsonMap;

import java.util.List;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class Color {

    @ColorInt
    public static final int TRANSPARENT = android.graphics.Color.TRANSPARENT;
    @ColorInt
    public static final int WHITE = android.graphics.Color.WHITE;
    @ColorInt
    public static final int BLACK = android.graphics.Color.BLACK;

    private final int defaultColor;
    @NonNull
    private final List<ColorSelector> selectors;

    public Color(int defaultColor, @NonNull List<ColorSelector> selectors) {
        this.defaultColor = defaultColor;
        this.selectors = selectors;
    }

    @NonNull
    public static Color fromJson(@NonNull JsonMap json) throws JsonException {
        JsonMap defaultColorJson = json.opt("default").optMap();
        @ColorInt Integer defaultColor = HexColor.fromJson(defaultColorJson);
        if (defaultColor == null) {
            throw new JsonException("Failed to parse color. 'default' may not be null! json = " + json);
        }
        JsonList selectorsJson = json.opt("selectors").optList();
        List<ColorSelector> selectors = ColorSelector.fromJsonList(selectorsJson);

        return new Color(defaultColor, selectors);
    }

    @Nullable
    public static Color fromJsonField(@Nullable JsonMap json, @NonNull String fieldName) throws JsonException {
        if (json == null || json.isEmpty()) {
            return null;
        }
        JsonMap colorJson = json.opt(fieldName).optMap();
        if (colorJson.isEmpty()) {
            return null;
        }
        return fromJson(colorJson);
    }

    public static float alpha(@ColorInt int color) {
        return android.graphics.Color.alpha(color);
    }

    @ColorInt
    public int resolve(@NonNull Context context) {
        // Look for a selector that matches the current UI mode.
        boolean isDarkMode = ResourceUtils.isUiModeNight(context);
        for (ColorSelector selector : selectors) {
            if (selector.isDarkMode() == isDarkMode) {
                return selector.getColor();
            }
        }
        // Fall back to default color if no match.
        return defaultColor;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/Color.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import android.content.Context
import android.graphics.Color
import androidx.annotation.ColorInt
import com.urbanairship.android.layout.util.ResourceUtils
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue
import com.urbanairship.json.optionalList
import com.urbanairship.json.optionalMap

public class Color public constructor(
    private val defaultColor: Int,
    private val selectors: List<ColorSelector>
) {

    @ColorInt
    public fun resolve(context: Context): Int {
        // Look for a selector that matches the current UI mode.
        val isDarkMode = ResourceUtils.isUiModeNight(context)
        return selectors
            .firstOrNull { it.isDarkMode == isDarkMode }
            ?.color
            ?: defaultColor
    }

    public companion object {

        @ColorInt
        public const val TRANSPARENT: Int = Color.TRANSPARENT

        @ColorInt
        public const val WHITE: Int = Color.WHITE

        @ColorInt
        public const val BLACK: Int = Color.BLACK

        private const val KEY_DEFAULT = "default"
        private const val KEY_SELECTORS = "selectors"

        @JvmStatic
        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): com.urbanairship.android.layout.property.Color {
            val content = json.requireMap()

            @ColorInt val defaultColor = HexColor.fromJson(content.optionalMap(KEY_DEFAULT))
            if (defaultColor == null) {
                throw JsonException("Failed to parse color. 'default' may not be null! json = $json")
            }

            val selectorsJson = content.opt(KEY_SELECTORS).optList()

            return Color(
                defaultColor = defaultColor,
                selectors = ColorSelector.fromJsonList(selectorsJson)
            )
        }

        @JvmStatic
        @Throws(JsonException::class)
        public fun fromJsonField(
            json: JsonMap?,
            fieldName: String
        ): com.urbanairship.android.layout.property.Color? {
            val source = json?.get(fieldName) ?: return null
            return fromJson(source)
        }

        @JvmStatic
        public fun alpha(@ColorInt color: Int): Float {
            return Color.alpha(color).toFloat()
        }
    }
}
